cachito/workers/pkg_managers/rubygems.py

```diff
@@ -136,3 +142,182 @@ def _validate_path_dependency_dir(gem, project_root, gemlock_dir):
+    :param GemMetadata gem: Gem dependency from a Gemfile.lock file
+    :param Path deps_dir: The deps/rubygems directory in a Cachito request bundle
+    :param str proxy_url: URL of Nexus RubyGems proxy
+    :param requests.auth.AuthBase proxy_auth: Authorization for the RubyGems proxy
+    """
+    package_dir = deps_dir / gem.name
+    package_dir.mkdir(exist_ok=True)
+    download_path = package_dir / f"{gem.name}-{gem.version}.gem"
+
+    proxied_url = f"{proxy_url.rstrip('/')}/gems/{gem.name}-{gem.version}.gem"
+    download_binary_file(proxied_url, download_path, auth=proxy_auth)
+
+    return {
+        "package": gem.name,
+        "version": gem.version,
+        "path": download_path,
+    }
+
+
+def _download_git_package(gem, rubygems_deps_dir, rubygems_raw_repo_name, nexus_auth):
+    """
+    Fetch the source for a Ruby package from Git.
+
+    If the package is already present in Nexus as a raw component, download it
+    from there instead of fetching from the original location.
+
+    :param GemMetadata gem: Git dependency from a Gemfile.lock file
+    :param Path rubygems_deps_dir: The deps/rubygems directory in a Cachito request bundle
+    :param str rubygems_raw_repo_name: Name of the Nexus raw repository for RubyGems
+    :param requests.auth.AuthBase nexus_auth: Authorization for the Nexus raw repo
+
+    :return: Dict with package name, download path, git url and ref, name of raw component in Nexus
+        and boolean whether we already have the raw component in Nexus
+    """
+    host, namespace_parts, repo_name = _extract_git_info(gem.source)
+
+    # Download to e.g. deps/rubygems/github.com/namespace/repo
+    package_dir = rubygems_deps_dir.joinpath(host, *namespace_parts, repo_name)
+    package_dir.mkdir(parents=True, exist_ok=True)
+
+    filename = f"{repo_name}-external-gitcommit-{gem.version}.tar.gz"
+    download_path = package_dir / filename
+    raw_component_name = f"{repo_name}/{filename}"
+
+    # Download raw component if we already have it
+    have_raw_component = _download_raw_component(
+        raw_component_name, rubygems_raw_repo_name, download_path, nexus_auth
+    )
+
+    if not have_raw_component:
+        log.debug("Raw component not found, will fetch from git")
+        repo_name = Git(gem.source, gem.version)
+        repo_name.fetch_source(gitsubmodule=False)
+        # Copy downloaded archive to expected download path
+        shutil.copy(repo_name.sources_dir.archive_path, download_path)
+
+    return {
+        "package": gem.name,
+        "path": download_path,
+        "url": gem.source,
+        "ref": gem.version.lower(),
+        "raw_component_name": raw_component_name,
+        "have_raw_component": have_raw_component,
+    }
+
+
+def _extract_git_info(repo_url):
+    """Extract git info from the repo url.
+
+    :param repo_url: url to the git repository
+    :return: host part of url, list with namespace parts and a repository name
+    """
```

Comment:
Would this also be a good candidate for refactoring into the `general` module?

Reply:
Yes. Function that pip uses does a little more than is needed for RubyGems, but to remove code duplication, it makes sense.